File: 2022/11/05/go/contains_duplicate.go
/**
 * https://leetcode.com/problems/contains-duplicate/
 * https://leetcode.com/problems/contains-duplicate-ii/
 * Author: Andrew Jarombek
 * Date: 11/5/2022
 */

package _go

func containsDuplicate(nums []int) bool {
	chars := make(map[int]struct{})

	for _, num := range nums {
		if _, ok := chars[num]; ok {
			return true
		}

		chars[num] = struct{}{}
	}

	return false
}

func containsNearbyDuplicate(nums []int, k int) bool {
	chars := make(map[int]struct{})

	for i, num := range nums {
		start := i - k - 1

		if start >= 0 {
			c := nums[start]
			delete(chars, c)
		}

		if _, ok := chars[num]; ok {
			return true
		}

		chars[num] = struct{}{}
	}

	return false
}

File: 2023/03/31/cpp/containsDuplicate.cpp
// https://leetcode.com/problems/contains-duplicate/
// Author: Andrew Jarombek
// Date: 3/31/2023

#include <cassert>
#include <vector>
#include <set>

using namespace std;

bool containsDuplicate(vector<int>& nums) {
    set<int> uniqueNums {};

    for (int num : nums) {
        if (uniqueNums.find(num) != uniqueNums.end()) {
            return true;
        }

        uniqueNums.insert(num);
    }

    return false;
}

int main() {
    vector<int> v {1, 2, 3};
    assert(!containsDuplicate(v));

    vector<int> v2 {1, 2, 1};
    assert(containsDuplicate(v2));
}